From: Markus Trippelsdorf <markus@trippelsdorf.de>
Date: Sun, 24 Jul 2011 14:03:30 +0200
Subject: xfs: Fix wrong return value of xfs_file_aio_write
References: bnc#719079
Patch-mainline: v3.1-rc1
Git-commit: 340a0a01b9675a16201cc4fc4a210eb5b3bc11ce

The fsync prototype change commit 02c24a82187d accidentally overwrote
the ssize_t return value of xfs_file_aio_write with 0 for SYNC type
writes. Fix this by checking if an error occured when calling
xfs_file_fsync and only change the return value in this case.
In addition xfs_file_fsync actually returns a normal negative error, so
fix this, too.

Signed-off-by: Markus Trippelsdorf <markus@trippelsdorf.de>
Reviewed-by: Christoph Hellwig <hch@lst.de>
Tested-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Acked-by: Jan Kara <jack@suse.cz>
Signed-off-by: Suresh Jayaraman <sjayaraman@suse.com>
---
 fs/xfs/linux-2.6/xfs_file.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

Index: linux-3.0-SLE11-SP2/fs/xfs/linux-2.6/xfs_file.c
===================================================================
--- linux-3.0-SLE11-SP2.orig/fs/xfs/linux-2.6/xfs_file.c
+++ linux-3.0-SLE11-SP2/fs/xfs/linux-2.6/xfs_file.c
@@ -900,11 +900,14 @@ xfs_file_aio_write(
 	/* Handle various SYNC-type writes */
 	if ((file->f_flags & O_DSYNC) || IS_SYNC(inode)) {
 		loff_t end = pos + ret - 1;
+		int error;
 
 		xfs_rw_iunlock(ip, iolock);
-		ret = -xfs_file_fsync(file, pos, end,
+		error = xfs_file_fsync(file, pos, end,
 				      (file->f_flags & __O_SYNC) ? 0 : 1);
 		xfs_rw_ilock(ip, iolock);
+		if (error)
+			ret = error;
 	}
 
 out_unlock:
